src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/AbstractTextDocumentPresentationEndpointBase.cs

```diff
@@ -0,0 +1,362 @@
+                return null;
+            }
+
+            var (documentSnapshot, version) = documentAndVersion;
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            if (request.Range.Start.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex) != true)
+            {
+                return null;
+            }
+
+            var languageKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            // See if we can handle this directly in Razor. If not, we'll let things flow to the below delegated handling.
+            var result = await TryGetRazorWorkspaceEditAsync(languageKind, request, cancellationToken).ConfigureAwait(false);
+            if (result is not null)
+            {
+                return result;
+            }
+
+            if (languageKind is not (RazorLanguageKind.CSharp or RazorLanguageKind.Html))
+            {
+                _logger.LogInformation($"Unsupported language {languageKind:G}.");
+                return null;
+            }
+
+            var requestParams = CreateRazorRequestParameters(request);
+
+            requestParams.HostDocumentVersion = version;
+            requestParams.Kind = languageKind;
+
+            // For CSharp we need to map the range to the generated document
+            if (languageKind == RazorLanguageKind.CSharp)
+            {
+                if (!_razorDocumentMappingService.TryMapToProjectedDocumentRange(codeDocument, request.Range, out var projectedRange))
+                {
+                    return null;
+                }
+
+                requestParams.Range = projectedRange;
+            }
+
+            var delegatedRequest = await _languageServer.SendRequestAsync(EndpointName, requestParams).ConfigureAwait(false);
+            var response = await delegatedRequest.Returning<WorkspaceEdit?>(cancellationToken).ConfigureAwait(false);
+            if (response is null)
+            {
+                return null;
+            }
+
+            // The responses we get back will be for virtual documents, so we have to map them back to the real
+            // document, and in the case of C#, map the returned ranges too
+            return MapWorkspaceEdit(response, mapRanges: languageKind == RazorLanguageKind.CSharp, codeDocument);
+        }
+
+        private static bool TryGetDocumentChanges(WorkspaceEdit workspaceEdit, [NotNullWhen(true)] out TextDocumentEdit[]? documentChanges)
```

Comment:
Love that you ported all of these. At some point I'll probably lift them onto the remapper too. ❤️ ❤️ ❤️

Reply:
There is definitely a bunch of servicification that would have helped here, but thats also a _huge_ job. eg, the remapping of workspace edits I ported from the LSP mapping provider whatsit, but that worked on VS types and I had to use O# types, so it was cool to port. Now my port is back to VS ports there is more overlap, but also that one makes LSP requests for the remapping, and goes down a different path, so I left mine. In an ideal world there would probably be some kind of mapping function that took a remapping provider or some such.. but then again if the HtmlCSharp server is going away, then that isn't needed. Its all a bit of a minefield if you ask me.